proposals/POEM-2-function-pulling-container-scheduler.md

```diff
@@ -0,0 +1,132 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+# Title
+Function Pulling Container Scheduler(FPCScheduler)
+
+## Status
+* Current state: Draft
+* Author(s): @style95, @ningyougang, @keonhee, @jiangpengcheng, @upgle
+
+## Summary and Motivation
+
+This POEM proposes a new scheduler for OpenWhisk, FPCScheduler.
+Previously we revealed [performance issues](https://cwiki.apache.org/confluence/display/OPENWHISK/Autonomous+Container+Scheduling+v1) in OpenWhisk.
+There are many reasons for them, but we can summarize them into the following problems.
+
+1. Multiple actions share the same `homeInvoker`.
+2. Current scheduler does not consider time taking for container operations(create/delete, pause/unpause).
+3. Resources are evenly divided by the number of controllers.
+
+First, multiple actions share the same `homeInvoker`. This is because the `homeInvoker` is statically decided by a hash function.
+While scheduling, resource status in the invoker side is not considered and this introduces busy hotspot invokers while the others are idle.
+This is generally a good idea to increase the probability of container reuse, but it becomes bad in the real scene because of slow container operation.
+When all resources in an invoker are taken by one action(busy/warm), an activation for another action will remove one of the existing containers and create a new one for it.
+If actions run for a long time this would be a good approach in terms of bin-packing. But when actions run for short time(most of the serverless cases),
+it severely increases the response time because container operations take one to two orders of magnitude longer time than invocation and it results in bad performance.
+
+Second, since controllers schedule activations to invokers in a distributed environment, controllers are assigned resource fraction of invokers.
+It's not feasible to collect the status of invokers, scheduling decisions made by all controllers and schedule based on them in real time so that controllers schedule activations within their own partial invoker resources.
+As a result, controllers throttle activations even if cluster-wide invocations are less than limits.
+
+We propose FPCScheduler to address the above issues.
+The scheduler differs from the existing scheduler(`ShardingPoolBalancer`) in many directions.
+It schedules containers rather than function requests. Each container is supposed to pull activations requests from the given action queue continuously.
+Whenever one execution is over, it fetches the next activation requests and invokes it repeatedly. In this way, we can maximize container reuse.
+One more benefit is schedulers don't need to track/consider the location of existing containers. Schedulers just decide where and when to create more containers.
```

Comment:
```suggestion
- An added benefit is that schedulers don't need to track or consider the location of existing containers. Schedulers decide where and when to create more containers.
```

Reply:
I don't understand this point. When/How is the decision to create more containers done?

Reply:
The scheduler component has multiple action queues.
Each action queue monitors its status such as the number of incoming requests, the number of requests in the queue, the number of requests that stayed in the queue for the given threshold(stale requests).
Based on this information, each queue decides the number of containers to be created.

There are many rules about this, but it generally works in this way:

1. If there is no container(when a queue is newly created), create one container immediately.
2. If there are stale requests in the queue which have stayed in the queue for some time without any processing, create more containers to handle them.
3. If there are more requests than the existing containers can handle, create more containers accordingly.

Reply:
How do the container proxies that are pulling these activations off the queue coordinate between invokers to determine if more containers need to be created? Seems like a hard problem to get the optimal number of containers up to keep up with requests without over provisioning

Reply:
It uses the average `duration` for the given(configurable) interval. So that we can calculate the number of activations that one container can handle. Then we can add more containers according to the factors such as the number of incoming activations, the number of activations in the queue, the number of containers which are being created, and so on. It's based on heuristics and there can be some differences at any time as it's difficult to accurately predict the required number of containers like what you said. Also, it tends to a little bit over-provisioning containers but they work with a priority queue which is based on container IDs. So if there are not enough activations to keep all containers up, surplus containers are removed by natural selection.